File: cpp/31_copy_list_random.cpp
/**
 * LeetCode #138: 随机链表的复制
 * 题目: 复制带 random 指针的链表。
 * 解法: 节点后插入复制节点 + 分离（三步）:
 *       1. 原链表每节点后插入复制节点（random 暂未设置）
 *       2. 遍历设置复制节点的 random（copy.random = origin.random.next）
 *       3. 分离两个链表
 * 时间 O(n)，空间 O(1)（不计结果空间）。
 */

#include <iostream>
#include <unordered_map>
using namespace std;

struct Node {
    int val;
    Node* next;
    Node* random;
    Node(int x) : val(x), next(nullptr), random(nullptr) {}
};

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (!head) return nullptr;
        // Step 1: 插入复制节点
        for (Node* cur = head; cur; ) {
            Node* copy = new Node(cur->val);
            copy->next = cur->next;
            cur->next = copy;
            cur = copy->next;
        }
        // Step 2: 设置 random
        for (Node* cur = head; cur; cur = cur->next->next)
            if (cur->random)
                cur->next->random = cur->random->next;
        // Step 3: 分离
        Node dummy(0);
        Node* tail = &dummy;
        for (Node* cur = head; cur; cur = cur->next) {
            tail->next = cur->next;
            tail = tail->next;
            cur->next = cur->next->next;
        }
        return dummy.next;
    }
};

int main() {
    Solution sol;
    Node* n1 = new Node(7);
    Node* n2 = new Node(13);
    Node* n3 = new Node(11);
    Node* n4 = new Node(10);
    Node* n5 = new Node(1);
    n1->next = n2; n2->next = n3; n3->next = n4; n4->next = n5;
    n2->random = n1; n3->random = n5; n4->random = n3; n5->random = n1;
    Node* copied = sol.copyRandomList(n1);
    cout << copied->val << " " << copied->next->val << endl; // 7 13
    return 0;
}

File: python/31_copy_list_random.py
#!/usr/bin/env python3
"""
LeetCode #138: 随机链表的复制
题目: 给定链表，每个节点有 next 和 random 指针，复制整个链表。
解法: 节点后插入复制节点 + 分离 - 三步：
      1. 在原链表每个节点后插入复制节点（random 暂未设置）
      2. 遍历设置复制节点的 random 指针（copy.random = origin.random.next）
      3. 分离两个链表
      时间 O(n)，空间 O(1)（不计结果空间）。
"""


class Node:
    def __init__(self, x, next=None, random=None):
        self.val = x
        self.next = next
        self.random = random


def copyRandomList(head: Node) -> Node:
    """
    三步复制法:
    1. 插入复制节点：origin.next = copy
    2. 设置复制节点的 random：copy.random = origin.random.next
    3. 分离：origin.next = copy.next, copy.next = origin.next.next
    """
    if not head:
        return None
    # Step 1: 插入复制节点
    cur = head
    while cur:
        copy = Node(cur.val)
        copy.next = cur.next
        cur.next = copy
        cur = copy.next
    # Step 2: 设置 random 指针
    cur = head
    while cur:
        if cur.random:
            cur.next.random = cur.random.next
        cur = cur.next.next
    # Step 3: 分离
    cur = head
    dummy = head.next
    while cur:
        copy = cur.next
        cur.next = copy.next
        if copy.next:
            copy.next = copy.next.next
        cur = cur.next
    return dummy


if __name__ == "__main__":
    # 测试：7 -> 13 -> 11 -> 10 -> 1，随机指针
    n1 = Node(7); n2 = Node(13); n3 = Node(11); n4 = Node(10); n5 = Node(1)
    n1.next = n2; n2.next = n3; n3.next = n4; n4.next = n5
    n1.random = None; n2.random = n1; n3.random = n5; n4.random = n3; n5.random = n1
    copied = copyRandomList(n1)
    print(copied.val, copied.next.val)  # 7, 13
